src/main/java/org/spongepowered/common/user/ServerUserProvider.java

```diff
@@ -0,0 +1,336 @@
+            this.watchKey = this.getSaveHandlerDirectory().register(
+                    this.filesystemWatchService,
+                    StandardWatchEventKinds.ENTRY_CREATE,
+                    StandardWatchEventKinds.ENTRY_DELETE);
+        } catch (final IOException e) {
+            SpongeCommon.getLogger().warn("Could not start file watcher");
+            if (this.filesystemWatchService != null) {
+                // it might be the watchKey that failed, so null it out again.
+                try {
+                    this.filesystemWatchService.close();
+                } catch (final IOException ex) {
+                    // ignored
+                }
+            }
+            this.watchKey = null;
+            this.filesystemWatchService = null;
+        }
+    }
+
+    void teardownWatchers() {
+        if (this.watchKey != null) {
+            this.watchKey.cancel();
+            this.watchKey = null;
+        }
+
+        if (this.filesystemWatchService != null) {
+            try {
+                this.filesystemWatchService.close();
+            } catch (final IOException e) {
+                // ignored - we're nulling this anyway
+            } finally {
+                this.filesystemWatchService = null;
+            }
+        }
+    }
+
+    void refreshFilesystemProfiles() {
+        if (this.watchKey != null && this.watchKey.isValid()) {
+            this.watchKey.reset();
+        }
+        this.knownUUIDs.clear();
+        this.userCache.invalidateAll();
+
+        // Add all known profiles from the data files
+        final String[] uuids = this.getSaveHandler().func_215771_d();
+        for (final String playerUuid : uuids) {
+
+            // If the filename contains a period, we can fail fast. Vanilla code fixes the Strings that have ".dat" to strip that out
+            // before passing that back in getAvailablePlayerDat. It doesn't remove non ".dat" filenames from the list.
+            if (playerUuid.contains(".")) {
+                continue;
+            }
+
+            // At this point, we have a filename who has no extension. This doesn't mean it is actually a UUID. We trap the exception and ignore
+            // any filenames that fail the UUID check.
+            final UUID uuid;
+            try {
+                uuid = UUID.fromString(playerUuid);
+            } catch (final Exception ex) {
+                continue;
+            }
+
+            this.knownUUIDs.add(uuid);
+        }
+    }
+
+    Optional<User> getUser(final String lastKnownName) {
+        final com.mojang.authlib.GameProfile gameProfile = this.server.getPlayerProfileCache().getGameProfileForUsername(lastKnownName);
+        if (gameProfile == null) {
+            return Optional.empty();
+        }
+        return this.getUser((GameProfile) gameProfile, true);
+    }
+
+    Optional<User> getUser(final UUID uuid) {
+        this.pollFilesystemWatcher();
```

Comment:
This method (getUser) is equivalent to `this.getUser(this.server.getPlayerProfileCache().getProfileByUUID(uuid), true);`